### PR TITLE
nanoBragg performance tuning

### PR DESCRIPTION
**Pull request by dwpaley**

### CODE REVIEW

**Comment by dwpaley** [comment]
@bkpoon Can you look at the new function in scitbx/math?

**Comment by dwpaley** [comment]
Rebased on master to get the xfel psana fix in 89fa3008

**Review by phyy-nx — APPROVED** [review]
I've reviewed the scitbx r3_rotation changes and they look good to me.

File: simtbx/nanoBragg/nanoBragg_crystal.py [hunk 1]
```diff
@@ -9,17 +9,20 @@
...
-    self.xtal_shape = "gauss"  # shapetype.Gauss
-    self.Ncells_abc = (10, 10, 10)
-    self.mos_spread_deg = 0
-    self.n_mos_domains = 1
-    self.thick_mm = 0.1
-    self.symbol = 'P43212'
-    self.miller_array = NBcrystal.dummie_Fhkl(ucell, self.symbol)
-    self.dxtbx_crystal = NBcrystal.dxtbx_crystal_from_ucell_and_symbol(ucell_tuple_Adeg=ucell,
-                                                                       symbol=(self.symbol))
+  def __init__(self, default=False):
```
Comment (dermen) [inline simtbx/nanoBragg/nanoBragg_crystal.py:12]: I thought python sometimes raises exceptions when creating objects whose properties are not initialized, at the least set them to None ? Some of the property methods might need to be updated, also can we keep the original behavior of defaulting to the default crystal   ?
```
def __init__(self, use_default_crystal=True): 
    self.xtal_shape = None  # string , one of the nanoBragg shape types, see below
    self.Ncells_abc = None  # 3 tuple defining mosaic domain size
    self.mos_spread_deg = None  # angular spread in degrees
    self.n_mos_domains = None  # number of domains 
    self.thick_mm = None  # crystal thickness used to scale the nanoBragg model 
    self.symbol = None  # space group symbol
    self.miller_array = None   # CCTBX miller array defining scattering amplitudes
    self.dxtbx_crystal = None     # dxtbx crystal model
    if use_default_crystal:
      ucell = (79.1, 79.1, 38.4, 90, 90, 90)
      self.xtal_shape = "gauss"  # shapetype.Gauss
      self.Ncells_abc = (10, 10, 10)
      self.mos_spread_deg = 0
      self.n_mos_domains = 1
      self.thick_mm = 0.1
      self.symbol = 'P43212'
      self.miller_array = NBcrystal.dummie_Fhkl(ucell, self.symbol)
      self.dxtbx_crystal = NBcrystal.dxtbx_crystal_from_ucell_and_symbol(ucell_tuple_Adeg=ucell,
                                                                         symbol=(self.symbol))
```

Also, I didnt think you would ever need to set self._miller_array directly  ?

File: simtbx/nanoBragg/nanoBragg_crystal.py [hunk 2]
```diff
@@ -9,17 +9,20 @@
...
+      self.Ncells_abc = (10, 10, 10)
+      self.mos_spread_deg = 0
+      self.n_mos_domains = 1
+      self.thick_mm = 0.1
+      self.symbol = 'P43212'
+      self.miller_array = NBcrystal.dummie_Fhkl(ucell, self.symbol)
+      self.dxtbx_crystal = NBcrystal.dxtbx_crystal_from_ucell_and_symbol(ucell_tuple_Adeg=ucell,
+                                                                         symbol=(self.symbol))
+    else:
+      self._miller_array = None
```
Comment (dermen) [inline simtbx/nanoBragg/nanoBragg_crystal.py:25]: revise the miller array setter

```
  @miller_array.setter
  def miller_array(self, val):
    if val is not None:
      if isinstance(val.data()[0], complex):
        self.miller_is_complex = True
      else:
        self.miller_is_complex = False
      if str(val.observation_type) == 'xray.intensity':
        val = val.as_amplitude_array()
      val = val.expand_to_p1()
      val = val.generate_bijvoet_mates()
    self._miller_array = val
```

**Comment by dwpaley** [comment]
Closing this since @nksauter is moving the code into master via individual commits 8e12c5ca6, 93f659c5, ee1208ee, and a couple more to come

**Pull request closed without merging**